### DISCUSSION TITLE
BasicDFN and DFN model

### DISCUSSION THREAD

**Discussion by xyyrly** [pybamm-team/PyBaMM#3202] [Q&A] [answered]
Hi Team,

 Does BasicDFN model still work now? I used BasicDFN and DFN to calculate the same event but got different results. For example, the "Negative electrode interfacial current density [A.m-2]" even has quite different trends:

![图片](https://github.com/pybamm-team/PyBaMM/assets/117139564/b1ea2b11-4726-4dd8-8195-c53dbeec091f)

**Comment by valentinsulzer** [comment]
That is strange, they are supposed to give identical answers, it is even tested here: https://github.com/pybamm-team/PyBaMM/blob/develop/tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_compare_basic_models.py#L13

**Reply by xyyrly** [reply]
I don't know why it doesn't work... Here is a simple example, and their variable "Negative electrode potential [V]" cannot match with each either.
model0= pybamm.lithium_ion.BasicDFN()
model_1=pybamm.lithium_ion.DFN()
parameter_value = pybamm.ParameterValues("Ecker2015")
parameter_value["Current function [A]"]=1.5652
simulation = pybamm.Simulation(model0,parameter_values=parameter_value)
simulation1 = pybamm.Simulation(model1,parameter_values=parameter_value)
t_eval = np.linspace(0, 0.005*5208, 5209)
solution_0 = simulation.solve(t_eval)
solution_0 = simulation.solution
solution_1 = simulation1.solve(t_eval)
solution_1 = simulation1.solution
plt.plot(solution_0["Negative electrode potential [V]"].entries[:,-1],label="BasicDFN")
plt.plot(solution_1["Negative electrode potential [V]"].entries[:,-1],label="DFN")
plt.xlabel("x/m")
plt.ylabel("Negative electrode potential [V]")
plt.legend()
![ded772f7e0315602f6c80ff6faf9d90](https://github.com/pybamm-team/PyBaMM/assets/117139564/cd017c4a-d0c5-42ce-9b81-3d94d6b40514)

**Reply by rtimms** [reply]
There is an error in the boundary conditions in `BasicDFN` which only shows up when the diffusivity is nonlinear. We should take `surf(D(c))` and not `D(surf(c))`.

The particle boundary conditions in `BasicDFN` should be replaced with 
```python3
        # Boundary conditions must be provided for equations with spatial derivatives
        self.boundary_conditions[c_s_n] = {
            "left": (pybamm.Scalar(0), "Neumann"),
            "right": (
                -j_n / (param.F * pybamm.surf(param.n.prim.D(c_s_n, T))),
                "Neumann",
            ),
        }
        self.boundary_conditions[c_s_p] = {
            "left": (pybamm.Scalar(0), "Neumann"),
            "right": (
                -j_p / (param.F * pybamm.surf(param.p.prim.D(c_s_p, T))),
                "Neumann",
            ),
        }
```

**Reply by rtimms** [reply, marked as answer]
Fixed here: https://github.com/pybamm-team/PyBaMM/pull/3207

**Reply by xyyrly** [reply]
Thank you very much. Now it works well.

**Reply by bobonice** [reply]
@rtimms Excellent job! After fixing this issue, another strange problem is that differences in the results of two models still persist. While this disparity may not be readily discernible in the figures, it becomes more evident when examining the numerical values. Additionally, the differences between the two gradually almost increase over time and distance, which could be a consequence of errors accumulating. And the other variables also exhibit the same issue, such as "Negative particle surface concentration [mol.m-3]". I would appreciate it if you could shed some light on the factors contributing to the differences in results between the BasicDFN and DFN models. 
![com1](https://github.com/pybamm-team/PyBaMM/assets/22030806/f2c4607a-9534-4e13-89b2-953f41c6475e)

**Reply by brosaplanella** [reply]
If you adjust the solver tolerances and/or refine the mesh, does the error decrease? It could be a matter of numerical accuracy. Here is an example on how to [change solver settings](https://docs.pybamm.org/en/latest/source/examples/notebooks/getting_started/tutorial-8-solver-options.html) and [change the mesh](https://docs.pybamm.org/en/latest/source/examples/notebooks/getting_started/tutorial-9-changing-the-mesh.html).

**Reply by valentinsulzer** [reply]
Alternatively can you check how the variance is when solving the same model twice

**Reply by bobonice** [reply]
Solving the same model twice, here is the test code and results. The difference of the same model is very small (~ 1e-6); however there are significant differences between different models (~1e-1). 
```c
import pybamm
import numpy as np
import matplotlib.pyplot as plt

model0= pybamm.lithium_ion.BasicDFN()
model1=pybamm.lithium_ion.DFN()
parameter_value = pybamm.ParameterValues("Ecker2015")
parameter_value["Current function [A]"]=1.5652

simulation00 = pybamm.Simulation(model0,parameter_values=parameter_value)
simulation01 = pybamm.Simulation(model0,parameter_values=parameter_value)
simulation10 = pybamm.Simulation(model1,parameter_values=parameter_value)
simulation11 = pybamm.Simulation(model1,parameter_values=parameter_value)

t_eval = np.linspace(0, 0.005*5208, 5209)
import datetime
print(datetime.datetime.now())
solution_00 = simulation00.solve(t_eval)
print(datetime.datetime.now())
solution_01 = simulation01.solve(t_eval)
print(datetime.datetime.now())
solution_10 = simulation10.solve(t_eval)
print(datetime.datetime.now())
solution_11 = simulation11.solve(t_eval)


plt.subplot(121)
plt.plot(np.abs(solution_00["Negative particle surface concentration [mol.m-3]"].entries[:,-1]
         -solution_01["Negative particle surface concentration [mol.m-3]"].entries[:,-1]),'r-',label="BasicDFN - BasicDFN")
plt.plot(np.abs(solution_10["Negative particle surface concentration [mol.m-3]"].entries[:,-1]
         -solution_11["Negative particle surface concentration [mol.m-3]"].entries[:,-1]),'k--',label="DFN - DFN")
plt.xlabel("x")
plt.ylabel("Negative particle surface concentration [mol.m-3]")
plt.legend()
plt.subplot(122)
plt.plot(np.abs(solution_00["Negative particle surface concentration [mol.m-3]"].entries[:,-1]
         -solution_10["Negative particle surface concentration [mol.m-3]"].entries[:,-1]),'r-',label="BasicDFN - DFN")
plt.plot(np.abs(solution_01["Negative particle surface concentration [mol.m-3]"].entries[:,-1]
         -solution_11["Negative particle surface concentration [mol.m-3]"].entries[:,-1]),'k--',label="BasicDFN - DFN")
plt.xlabel("x")
plt.ylabel("Negative particle surface concentration [mol.m-3]")
plt.legend()
plt.show()
```
![Figure_1](https://github.com/pybamm-team/PyBaMM/assets/22030806/0e3fe8bf-44da-4be1-a3d1-c9a789e79eef)

**Reply by rtimms** [reply]
The difference is still very small in relative terms - e.g. for the concentrations the solver is calculating error based on the dimensionless concentration (i.e. the concentration scaled by the "maximum concentration ..." parameter). My guess is because the model definition/formulation is slightly different then the solver error is calculated differently. I would plot the difference in dimensionless concentration between the basic DFN and DFN, looping over reducing solver tolerances. What I suspect you will see is that the difference in the solutions is about the tolerance you pass to the solver.
